# STAT 207 Lab Assignment 12 - Group Part - [9 Points]



## <u>Case Study</u>: Predicting Airbnb Superhost Status

Suppose that you have a spare room in your house (in Seattle) and you are considering advertising this room to guests on Airbnb. Airbnb superhosts are considered to be experienced hosts who provide a shining example for other hosts, and extraordinary experiences for their guests. Once a host reaches Superhost status, a superhost badge will automatically appear on their listing and profile to help guests identify them. You would like to assess your chances of being named a superhost with your property.

The following dataset is a sample (assume random) of available Airbnb listings in Seattle, WA. These listings were collected in January 2016, and filtered to just contain listings from the five most popular Seattlbe neighborhoods (for Airbnb listings) and just contain listings that are either in a house or apartment property.

This dataset contains the following variables.

**Listing Information**
The dataset contains the following information about the Airbnb *listing*:
* <u>price</u>: price of the listing (per night)(in US dollars)
* <u>review_scores_rating</u>: the average rating of the listing [0,100] (100 is the best)
* <u>number_of_reviews</u>: the number of reviews for the listing
* <u>security_deposit</u>: the security deposit required for the listing (in US dollars)
* <u>cleaning_fee</u>: the cleaning fee required for the listing (in US dollars)
* <u>neighborhood</u>: the neighborhood of Seattle the listing is located in
* <u>property_type</u>: is the listing in a 'House' or 'Apartment'
* <u>room_type</u>: is the listing a 'Entire home/apt', 'Private room', or 'Shared room'
* <u>accommodates</u>: how many guests will the listing accommodate
* <u>bathrooms</u>:how many bathrooms does the listing have
* <u>beds</u>: how many beds does the listing have

**Host Information**
The dataset also contains the following information about the *host* of the given Airbnb listing:
* <u>host_is_superhost</u>: is the host a "superhost": t=True, f=False
* <u>host_has_profile_pic</u>: does the host have a profile pic in their bio: t=True, f=False
* <u>host_response_time</u>: how fast will the host respond to requests (on average)

## Research Goals

The primary research goal that we will pursue in this group lab will be to build a predictive model that effectively predicts the `probability a Seattle Airbnb host is a **superhost**` for ***LISTINGS IN THIS DATASET*** given some combination of the available explanatory variables.



## Points

<p>&nbsp;</p>
<table style="border: none;border-collapse: collapse;width:102pt;">
    <tbody>
        <tr>
            <td style="padding:0px;color:black;font-size:15px;font-weight:700;font-style:normal;text-decoration:none;font-family:Calibri, sans-serif;text-align:right;vertical-align:bottom;border:.5pt solid windowtext;height:14.25pt;width:51pt;">Problem</td>
            <td style="padding:0px;color:black;font-size:15px;font-weight:700;font-style:normal;text-decoration:none;font-family:Calibri, sans-serif;text-align:general;vertical-align:bottom;border:.5pt solid windowtext;border-left:none;width:51pt;">Points</td>
        </tr>
        <tr>
            <td style="padding:0px;color:black;font-size:15px;font-weight:400;font-style:normal;text-decoration:none;font-family:Calibri, sans-serif;text-align:right;vertical-align:bottom;border:.5pt solid windowtext;height:14.25pt;border-top:none;">1.1</td>
            <td style="padding:0px;color:black;font-size:15px;font-weight:400;font-style:normal;text-decoration:none;font-family:Calibri, sans-serif;text-align:right;vertical-align:bottom;border:.5pt solid windowtext;border-top:none;border-left:none;">0.25</td>
        </tr>
        <tr>
            <td style="padding:0px;color:black;font-size:15px;font-weight:400;font-style:normal;text-decoration:none;font-family:Calibri, sans-serif;text-align:right;vertical-align:bottom;border:.5pt solid windowtext;height:14.25pt;border-top:none;">1.2</td>
            <td style="padding:0px;color:black;font-size:15px;font-weight:400;font-style:normal;text-decoration:none;font-family:Calibri, sans-serif;text-align:right;vertical-align:bottom;border:.5pt solid windowtext;border-top:none;border-left:none;">0.25</td>
        </tr>
        <tr>
            <td style="padding:0px;color:black;font-size:15px;font-weight:400;font-style:normal;text-decoration:none;font-family:Calibri, sans-serif;text-align:right;vertical-align:bottom;border:.5pt solid windowtext;height:14.25pt;border-top:none;">2.1.1</td>
            <td style="padding:0px;color:black;font-size:15px;font-weight:400;font-style:normal;text-decoration:none;font-family:Calibri, sans-serif;text-align:right;vertical-align:bottom;border:.5pt solid windowtext;border-top:none;border-left:none;">0.75</td>
        </tr>
        <tr>
            <td style="padding:0px;color:black;font-size:15px;font-weight:400;font-style:normal;text-decoration:none;font-family:Calibri, sans-serif;text-align:right;vertical-align:bottom;border:.5pt solid windowtext;height:14.25pt;border-top:none;">2.1.2</td>
            <td style="padding:0px;color:black;font-size:15px;font-weight:400;font-style:normal;text-decoration:none;font-family:Calibri, sans-serif;text-align:right;vertical-align:bottom;border:.5pt solid windowtext;border-top:none;border-left:none;">0.75</td>
        </tr>
        <tr>
            <td style="padding:0px;color:black;font-size:15px;font-weight:400;font-style:normal;text-decoration:none;font-family:Calibri, sans-serif;text-align:right;vertical-align:bottom;border:.5pt solid windowtext;height:14.25pt;border-top:none;">2.1.3</td>
            <td style="padding:0px;color:black;font-size:15px;font-weight:400;font-style:normal;text-decoration:none;font-family:Calibri, sans-serif;text-align:right;vertical-align:bottom;border:.5pt solid windowtext;border-top:none;border-left:none;">0.25</td>
        </tr>
        <tr>
            <td style="padding:0px;color:black;font-size:15px;font-weight:400;font-style:normal;text-decoration:none;font-family:Calibri, sans-serif;text-align:right;vertical-align:bottom;border:.5pt solid windowtext;height:14.25pt;border-top:none;">2.1.4</td>
            <td style="padding:0px;color:black;font-size:15px;font-weight:400;font-style:normal;text-decoration:none;font-family:Calibri, sans-serif;text-align:right;vertical-align:bottom;border:.5pt solid windowtext;border-top:none;border-left:none;">0.25</td>
        </tr>
        <tr>
            <td style="padding:0px;color:black;font-size:15px;font-weight:400;font-style:normal;text-decoration:none;font-family:Calibri, sans-serif;text-align:right;vertical-align:bottom;border:.5pt solid windowtext;height:14.25pt;border-top:none;">2.1.5</td>
            <td style="padding:0px;color:black;font-size:15px;font-weight:400;font-style:normal;text-decoration:none;font-family:Calibri, sans-serif;text-align:right;vertical-align:bottom;border:.5pt solid windowtext;border-top:none;border-left:none;">0.25</td>
        </tr>
        <tr>
            <td style="padding:0px;color:black;font-size:15px;font-weight:400;font-style:normal;text-decoration:none;font-family:Calibri, sans-serif;text-align:right;vertical-align:bottom;border:.5pt solid windowtext;height:14.25pt;border-top:none;">3.1</td>
            <td style="padding:0px;color:black;font-size:15px;font-weight:400;font-style:normal;text-decoration:none;font-family:Calibri, sans-serif;text-align:right;vertical-align:bottom;border:.5pt solid windowtext;border-top:none;border-left:none;">0.5</td>
        </tr>
        <tr>
            <td style="padding:0px;color:black;font-size:15px;font-weight:400;font-style:normal;text-decoration:none;font-family:Calibri, sans-serif;text-align:right;vertical-align:bottom;border:.5pt solid windowtext;height:14.25pt;border-top:none;">3.2</td>
            <td style="padding:0px;color:black;font-size:15px;font-weight:400;font-style:normal;text-decoration:none;font-family:Calibri, sans-serif;text-align:right;vertical-align:bottom;border:.5pt solid windowtext;border-top:none;border-left:none;">0.75</td>
        </tr>
        <tr>
            <td style="padding:0px;color:black;font-size:15px;font-weight:400;font-style:normal;text-decoration:none;font-family:Calibri, sans-serif;text-align:right;vertical-align:bottom;border:.5pt solid windowtext;height:14.25pt;border-top:none;">3.3</td>
            <td style="padding:0px;color:black;font-size:15px;font-weight:400;font-style:normal;text-decoration:none;font-family:Calibri, sans-serif;text-align:right;vertical-align:bottom;border:.5pt solid windowtext;border-top:none;border-left:none;">1</td>
        </tr>
        <tr>
            <td style="padding:0px;color:black;font-size:15px;font-weight:400;font-style:normal;text-decoration:none;font-family:Calibri, sans-serif;text-align:right;vertical-align:bottom;border:.5pt solid windowtext;height:14.25pt;border-top:none;">4.1</td>
            <td style="padding:0px;color:black;font-size:15px;font-weight:400;font-style:normal;text-decoration:none;font-family:Calibri, sans-serif;text-align:right;vertical-align:bottom;border:.5pt solid windowtext;border-top:none;border-left:none;">0.75</td>
        </tr>
        <tr>
            <td style="padding:0px;color:black;font-size:15px;font-weight:400;font-style:normal;text-decoration:none;font-family:Calibri, sans-serif;text-align:right;vertical-align:bottom;border:.5pt solid windowtext;height:14.25pt;border-top:none;">4.2</td>
            <td style="padding:0px;color:black;font-size:15px;font-weight:400;font-style:normal;text-decoration:none;font-family:Calibri, sans-serif;text-align:right;vertical-align:bottom;border:.5pt solid windowtext;border-top:none;border-left:none;">1</td>
        </tr>
        <tr>
            <td style="padding:0px;color:black;font-size:15px;font-weight:400;font-style:normal;text-decoration:none;font-family:Calibri, sans-serif;text-align:right;vertical-align:bottom;border:.5pt solid windowtext;height:14.25pt;border-top:none;">5.1</td>
            <td style="padding:0px;color:black;font-size:15px;font-weight:400;font-style:normal;text-decoration:none;font-family:Calibri, sans-serif;text-align:right;vertical-align:bottom;border:.5pt solid windowtext;border-top:none;border-left:none;">1</td>
        </tr>
        <tr>
            <td style="padding:0px;color:black;font-size:15px;font-weight:400;font-style:normal;text-decoration:none;font-family:Calibri, sans-serif;text-align:right;vertical-align:bottom;border:.5pt solid windowtext;height:14.25pt;border-top:none;">5.2</td>
            <td style="padding:0px;color:black;font-size:15px;font-weight:400;font-style:normal;text-decoration:none;font-family:Calibri, sans-serif;text-align:right;vertical-align:bottom;border:.5pt solid windowtext;border-top:none;border-left:none;">0.5</td>
        </tr>
        <tr>
            <td style="padding:0px;color:black;font-size:15px;font-weight:400;font-style:normal;text-decoration:none;font-family:Calibri, sans-serif;text-align:right;vertical-align:bottom;border:.5pt solid windowtext;height:14.25pt;border-top:none;">5.3</td>
            <td style="padding:0px;color:black;font-size:15px;font-weight:400;font-style:normal;text-decoration:none;font-family:Calibri, sans-serif;text-align:right;vertical-align:bottom;border:.5pt solid windowtext;border-top:none;border-left:none;">0.75</td>
        </tr>
    </tbody>
</table>

## 1. Basic Data Examination and Manipulation

### 1.1. Basics
1. Read this csv file into a dataframe.

2. Display your dataframe.

3. Show how many rows your dataframe has.



In [50]:
import pandas as pd
import numpy as np

In [51]:
airbnb = pd.read_csv('seattle_airbnb_listings_cleaned.csv')

# Displaying the first few rows of the DataFrame

airbnb_head = airbnb.head()



# Showing the number of rows in the DataFrame

num_rows = airbnb.shape[0]



airbnb_head, num_rows

(   price  review_scores_rating  number_of_reviews  security_deposit  \
 0    300                   100                 24               500   
 1    149                    96                 11               300   
 2     95                    95                 79               150   
 3    105                   100                 13               500   
 4    140                    99                 30               250   
 
    cleaning_fee neighborhood property_type        room_type  accommodates  \
 0            95  Wallingford         House  Entire home/apt             5   
 1           105  Wallingford     Apartment  Entire home/apt             6   
 2            40  Wallingford     Apartment  Entire home/apt             3   
 3            50  Wallingford         House     Private room             2   
 4            65  Wallingford         House  Entire home/apt             2   
 
    bathrooms  beds  host_response_time  host_response_rate host_is_superhost  \
 0        1.5  

### 1.2. 0/1 Response Variable

Create a new 0/1 variable in your dataset in which `superhosts=1` and `nonsuperhosts=0`. 

In [52]:
airbnb['host_is_superhost'] = airbnb['host_is_superhost'].replace({'t':1, 'f':0})
airbnb.head()

price  review_scores_rating  number_of_reviews  security_deposit  \
0    300                   100                 24               500   
1    149                    96                 11               300   
2     95                    95                 79               150   
3    105                   100                 13               500   
4    140                    99                 30               250   

   cleaning_fee neighborhood property_type        room_type  accommodates  \
0            95  Wallingford         House  Entire home/apt             5   
1           105  Wallingford     Apartment  Entire home/apt             6   
2            40  Wallingford     Apartment  Entire home/apt             3   
3            50  Wallingford         House     Private room             2   
4            65  Wallingford         House  Entire home/apt             2   

   bathrooms  beds  host_response_time  host_response_rate  host_is_superhost  \
0        1.5     3  within a few hours                 1.0                  1   
1        1.0     3      within an hour                 1.0                  0   
2        1.0     2      within an hour                 1.0                  0   
3        2.0     1  within a few hours                 1.0                  1   
4        1.0     1      within an hour                 1.0                  1   

  host_has_profile_pic host_identity_verified  
0                    t                      t  
1                    t                      t  
2                    t                      t  
3                    t                      t  
4                    t                      t

## 2. Full Model Logistic Regression with Scikit Learn

Next, rather than use the `smf.logit()` function to fit a logistic regression model as we've discussed in class, let's learn how to use the `LogisticRegression()` function from scikit learn to fit a logistic regression model instead.




In [53]:
from sklearn.linear_model import LogisticRegression

### 2.1. Full Logistic Regression Model

In the interest of time, we'll call our "full" model the one that predicts the `probability that a host is a superhost` given the four explanatory variables below:
* `review_scores_rating`
* `number_of_reviews`
* `neighborhood`
* `room_type`


#### 2.1.1. Features Matrix and Target Array

Just like the `LinearRegression()` function, the `LogisticRegression()` requires us to create a features matrix (of just the explanatory variables we intend to use) and a target array (ie. our 0/1 response variable).

Create your features matrix and target array below for this full model.

In [54]:
X_train = airbnb[['review_scores_rating','number_of_reviews','neighborhood', 'room_type']]
X_train

review_scores_rating  number_of_reviews neighborhood        room_type
0                     100                 24  Wallingford  Entire home/apt
1                      96                 11  Wallingford  Entire home/apt
2                      95                 79  Wallingford  Entire home/apt
3                     100                 13  Wallingford     Private room
4                      99                 30  Wallingford  Entire home/apt
..                    ...                ...          ...              ...
334                    95                  4      Fremont  Entire home/apt
335                    95                 12      Fremont  Entire home/apt
336                    98                 10      Fremont  Entire home/apt
337                    96                 29      Fremont  Entire home/apt
338                    92                  5      Fremont  Entire home/apt

[339 rows x 4 columns]

In [55]:
Y_train = airbnb['host_is_superhost']
Y_train

0      1
1      0
2      0
3      1
4      1
      ..
334    0
335    1
336    0
337    0
338    0
Name: host_is_superhost, Length: 339, dtype: int64

#### 2.1.2. Indicator Variables

Furthermore, we need to translate our categorical explanatory variables into indicator variables in this features matrix. Do so below.

In [56]:
X_train_encoded = pd.get_dummies(X_train,drop_first=True, dtype=int)
X_train_encoded

review_scores_rating  number_of_reviews  neighborhood_Broadway  \
0                     100                 24                      0   
1                      96                 11                      0   
2                      95                 79                      0   
3                     100                 13                      0   
4                      99                 30                      0   
..                    ...                ...                    ...   
334                    95                  4                      0   
335                    95                 12                      0   
336                    98                 10                      0   
337                    96                 29                      0   
338                    92                  5                      0   

     neighborhood_Fremont  neighborhood_Minor  neighborhood_Wallingford  \
0                       0                   0                         1   
1                       0                   0                         1   
2                       0                   0                         1   
3                       0                   0                         1   
4                       0                   0                         1   
..                    ...                 ...                       ...   
334                     1                   0                         0   
335                     1                   0                         0   
336                     1                   0                         0   
337                     1                   0                         0   
338                     1                   0                         0   

     room_type_Private room  room_type_Shared room  
0                         0                      0  
1                         0                      0  
2                         0                      0  
3                         1                      0  
4                         0                      0  
..                      ...                    ...  
334                       0                      0  
335                       0                      0  
336                       0                      0  
337                       0                      0  
338                       0                      0  

[339 rows x 8 columns]

#### 2.1.3. Instantiating a Logistic Regression Model

Now, instantiate a logistic regression model using the code below. In the code below we happen to call this instantiated model `full_mod`.

`full_mod = LogisticRegression()`

In [57]:
full_mod = LogisticRegression()

#### 2.1.4. Fitting the Instantiated Logistic Regression Model

Much like what we did with an instantiated LinearRegression() object, we can fit an instantiated LogisticRegression() object with the following code below.

`full_mod.fit(YOUR_FEATURES_MATRIX,YOUR_TARGET_ARRAY)`

Fit your logistic regression model with your features matrix and target array below.

In [58]:
full_mod.fit(X_train_encoded,Y_train)

/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/sklearn/linear_model/_logistic.py:458: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


LogisticRegression()

#### 2.1.5. Predictive Probabilities

Now that we've fit our model, let's predict the probability that each of our hosts in the features matrix is a superhost. You can do so with the code below.

`full_mod.predict_proba(YOUR_FEATURES_MATRIX)[:,1]`

Calculate these model predictive probabilities for the observations in your features matrix.

In [59]:
full_mod.predict_proba(X_train_encoded)[:,1]

array([5.07197527e-01, 2.73919129e-01, 4.67005888e-01, 4.35035703e-01,
       4.79975357e-01, 3.25686382e-01, 2.99099726e-02, 2.55989602e-01,
       3.74517585e-01, 5.85434617e-01, 2.86297371e-01, 5.36706763e-01,
       2.67856518e-01, 1.28094112e-02, 1.73281059e-01, 4.68121659e-01,
       3.11387580e-01, 2.70877163e-01, 2.38575820e-01, 6.19927514e-01,
       4.28378775e-02, 1.44781795e-01, 5.26412122e-02, 3.02868500e-03,
       2.35260322e-01, 2.25103412e-02, 2.99318193e-01, 4.28378775e-02,
       4.34669321e-01, 4.15967472e-02, 2.38575820e-01, 2.99318193e-01,
       6.44584933e-02, 8.95532167e-02, 4.23390554e-01, 1.28094112e-02,
       7.27681104e-03, 2.79162719e-01, 5.02987029e-01, 1.58748886e-02,
       2.16990765e-01, 4.78855592e-01, 3.64141448e-01, 5.26412122e-02,
       4.30168266e-01, 1.64866961e-01, 1.84551685e-01, 1.28094112e-02,
       6.09279497e-02, 1.56390167e-01, 3.85010363e-01, 5.32888297e-01,
       4.72690414e-01, 6.85542676e-01, 3.02861698e-01, 4.87641300e-01,
      

## 3. k-Fold Cross-Validation with the Full Logistic Regression Model

Next, let's learn how to perform k-fold cross-validation on our full logistic regression model.

### 3.1. Instantiate a Logistic Regression Model

First, instantiate (but don't fit) another logistic regression model object like we did in #2.1.3.

In [60]:
full_mod2 = LogisticRegression()

### 3.2. k=5 Fold Cross-Validation

Next, instantiate a k=5 fold object that *will* randomly split the dataset up into k=5 test folds. Your dataset should be randomly shuffled first before creating the test folds. Use a random state of 100.

In [61]:
from sklearn.model_selection import KFold
df_K5 = KFold(n_splits = 5, shuffle=True, random_state=100)

### 3.3. k-Fold Cross-Validation

1. Finally, use these two objects that you created in 3.1 and 3.2 to perform k=5 fold cross-validation on your **full features matrix** and your target array. You should use the **AUC of the test ROC curve** to evaluate each of your 5 training dataset model's corresponding test datasets.
2. Display each of the 5 test AUC values.
3. Display the mean of these 5 test AUC values.
4. Display the standard deviation of these 5 test AUC values.

**Hint:** *You can use the `cross_val_score()` function to perform k-fold cross-validation on a logistic regression model in the same way that we used it to perform k-fold cross-validation on a linear regression model. The only thing that you would want to change is the parameter `scoring="roc_auc"` in the function. This will calculate the AUC of each of the test dataset ROC curves.*

In [62]:
from sklearn.model_selection import cross_val_score
auc = cross_val_score(full_mod, X_train_encoded, Y_train, cv=df_K5, scoring="roc_auc" )
auc, auc.mean(), auc.std()


/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/sklearn/linear_model/_logistic.py:458: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(
/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/sklearn/linear_model/_logistic.py:458: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://sciki

(array([0.81023256, 0.81352657, 0.6875    , 0.72777778, 0.80592105]),
 0.7689915917194401,
 0.051744419351004954)

## 4. Cross-Validation for a Reduced Model

Now, we would like to perform k=5 fold cross-validation on a *reduced* logistic regression model. This reduced model will predict the `probability that a host is a superhost` given the three explanatory variables below:
* `review_scores_rating`
* `number_of_reviews`
* <strike>`neighborhood`</strike>
* `room_type`

### 4.1. Reduced Features Matrix

First create another features matrix that only corresponds to the three explanatory variables that we will use in the reduced model (ie. `review_scores_rating`, `number_of_reviews`, and `room_type`.) This features matrix should still use the nessary indicator variables.

In [63]:
airbnb_new = pd.get_dummies(x_train[["review_scores_rating", "number_of_reviews", "room_type"]],drop_first=True, dtype=int)
airbnb_new


review_scores_rating  number_of_reviews  room_type_Private room  \
0                     100                 24                       0   
1                      96                 11                       0   
2                      95                 79                       0   
3                     100                 13                       1   
4                      99                 30                       0   
..                    ...                ...                     ...   
334                    95                  4                       0   
335                    95                 12                       0   
336                    98                 10                       0   
337                    96                 29                       0   
338                    92                  5                       0   

     room_type_Shared room  
0                        0  
1                        0  
2                        0  
3                        0  
4                        0  
..                     ...  
334                      0  
335                      0  
336                      0  
337                      0  
338                      0  

[339 rows x 4 columns]

### 4.2. k-Fold Cross-Validation

1. Finally, use these two objects that you created in 3.1 and 3.2 to perform k=5 fold cross-validation on your **reduced features matrix** (that you created in #4.1) and your target array. You should use the **AUC of the test ROC curve** to evaluate each of your 5 training dataset model's corresponding test datasets.
2. Display each of the 5 test AUC values.
3. Display the mean of these 5 test AUC values.
4. Display the standard deviation of these 5 test AUC values.

In [64]:
from sklearn.model_selection import cross_val_score
auc2 = cross_val_score(full_mod, airbnb_new, Y_train, cv=df_K5, scoring="roc_auc" )
auc2, auc2.mean(), auc2.std()

(array([0.78744186, 0.76908213, 0.68629808, 0.76833333, 0.83826754]),
 0.7698845880370081,
 0.048937187576459075)

## 5. Model Evaluation

Based on the k-fold cross-validation AUC results for both the full and reduced logistic regression models:

1. **Full Model Results**:
   - AUC Values: [$0.8102, 0.8135, 0.6875, 0.7278, 0.8059$]
   - Mean AUC: $0.7690$
   - Standard Deviation: $0.0517$

2. **Reduced Model Results**:
   - AUC Values: [$0.7874, 0.7691, 0.6863, 0.7683, 0.8383$]
   - Mean AUC: $0.7699$
   - Standard Deviation: $0.0489$


### 5.1. Overall Best Cross-Validation Performance

Which of these two logistic regression models that we fitted above had the best k-fold cross-validation performance?

The comparative analysis of the full and reduced logistic regression models, based on their k-fold cross-validation AUC values, indicates that the reduced model performs marginally better. The mean AUC for the full model is 0.7690, while the reduced model has a slightly higher mean AUC of 0.7699. Although the difference is minimal, the reduced model, which excludes the `neighborhood` variable, shows a marginally superior performance in terms of the average AUC.

### 5.2. Relative Cross-Validation Performance

Would you say that the model that you selected in 5.1 has *comparatively* (or significantly) better performance than the other model? Explain.

Despite the reduced model having a marginally higher mean AUC, the difference in performance between the full and reduced models is not substantial. Both models exhibit similar mean AUCs (0.7690 for the full model and 0.7699 for the reduced model) and comparable standard deviations (0.0517 for the full model and 0.0489 for the reduced model). This suggests that the removal of the `neighborhood` variable does not lead to a significant improvement in model performance. Therefore, it can be concluded that both models perform comparably in predicting the superhost status.


### 5.3. Overfitting

Does your analysis above suggest that the neighborhood variable was overfitting the logistic regression model? Explain.

The analysis does not provide strong evidence that the `neighborhood` variable was causing overfitting in the full logistic regression model. The slight improvement in the AUC for the reduced model suggests that while the `neighborhood` variable may not be critically important for the model, its presence does not substantially degrade the model's ability to generalize to new data. Generally, overfitting is more likely when there are a large number of predictors, especially if they are not strongly relevant to the outcome. In this case, the `neighborhood` variable's influence appears to be limited but not detrimental to the model's overall predictive power.

Sophie Bustos and I worked together throughout the entire lab.